### PR TITLE
[Network] NAT Gateway support

### LINKED ISSUE #9090: NSG support for additional protocols
**Resource Provider**
Microsoft.Network

**Description of Feature or Work Requested**
Accepting new protocol on NSG security rules: 'Icmp'

**Minimum API Version Required**
2019-02-01

**Swagger Link**
https://github.com/Azure/azure-rest-api-specs/pull/5582/files
Note: 'Esp', 'Ah' are not ready yet, only ICMP for now

**Target Date**
May release

### LINKED ISSUE #9388: Profile 2018-03-01: eventhub module is not able to load
ADO run log:
    https://dev.azure.com/azure-sdk/public/_build/results?buildId=30908

Run task for `Profile 2018-03-01` contains lots of error:
     ERROR: Error loading command module 'eventhubs'

**Debug Log:**

> Error loading command module 'eventhubs'
25444 : 2019-05-09 17:38:05,634 : DEBUG : cli.azure.cli.core : Traceback (most recent call last):
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\profiles\_shared.py", line 224, in get_api_version
    api_version = AZURE_API_PROFILES[api_profile][resource_type]
KeyError: <ResourceType.MGMT_EVENTHUB: ('azure.mgmt.eventhub', 'EventHubManagementClient')>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\__init__.py", line 144, in _update_command_table_from_modules
    module_command_table, module_group_table = _load_module_command_loader(self, args, mod)
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\commands\__init__.py", line 893, in _load_module_command_loader
    return _load_command_loader(loader, args, mod, 'azure.cli.command_modules.')
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\commands\__init__.py", line 875, in _load_command_loader
    command_table = command_loader.load_command_table(args)
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\command_modules\eventhubs\__init__.py", line 29, in load_command_table
    load_command_table(self, args)
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\command_modules\eventhubs\commands.py", line 104, in load_command_table
    g.custom_command('add', 'cli_networkrule_createupdate', validator=validate_subnet)
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\commands\__init__.py", line 1073, in custom_command
    return self._command(name, method_name=method_name, custom_command=True, **kwargs)
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\commands\__init__.py", line 1085, in _command
    **merged_kwargs)
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\__init__.py", line 472, in _cli_command
    operation_group=kwargs.get('operation_group')):
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\__init__.py", line 391, in supported_api_version
    operation_group=operation_group)
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\profiles\__init__.py", line 44, in supported_api_version
    operation_group=operation_group)
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\profiles\_shared.py", line 354, in supported_api_version
    if isinstance(resource_type, (ResourceType, CustomResourceType)) else api_profile
  File "C:\Program Files (x86)\Python35\lib\site-packages\azure\cli\core\profiles\_shared.py", line 231, in get_api_version
    raise APIVersionException(resource_type, api_profile)
azure.cli.core.profiles._shared.APIVersionException: Unable to get API version for type 'ResourceType.MGMT_EVENTHUB' in profile '2018-03-01-hybrid'

### PR DESCRIPTION
**Pull request by tjprescott**
Closes #9090.
Closes #9388.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by limingu** [comment]
Azure.azure-cli (Build Homebrew Formula) Failing after 3m — Build Homebrew Formula failed

File: src/command_modules/azure-cli-natgateway/azure/cli/command_modules/natgateway/custom.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+def create_nat_gateway(cmd, nat_gateway_name, resource_group_name,
+                       location=None, public_ip_addresses=None,
+                       public_ip_prefixes=None, idle_timeout=None, no_wait=False):
+
+    if not public_ip_addresses and not public_ip_prefixes:
+        raise CLIError('usage error: --public-ip-addresses ADDRESSES | --public-ip-prefixes PREFIXES')
+
+    if public_ip_addresses is not None and public_ip_prefixes is not None:
+        logger.error("public_ip_addresses OR public_ip_prefixes required")
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-natgateway/azure/cli/command_modules/natgateway/custom.py]: Why not throw here?

Comment (limingu) [inline src/command_modules/azure-cli-natgateway/azure/cli/command_modules/natgateway/custom.py]: Why not raise error? If both public ip and public ip prefixes passed in, which one will take high priority?

File: src/command_modules/azure-cli-natgateway/azure/cli/command_modules/natgateway/validators.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+# pylint: disable=too-many-lines
+
+from azure.cli.core.commands.client_factory import get_subscription_id
+
+
+def validate_public_ip_addresses(resource):
+    from msrestazure.tools import is_valid_resource_id, resource_id
+
+    def _validate_public_ip_addresses(cmd, namespace):
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-natgateway/azure/cli/command_modules/natgateway/validators.py]: I saw a new file above named as `_validators.py`, but contents look different. Can you clarify?

**Comment by tjprescott** [comment]
Comments addressed.

File: src/command_modules/azure-cli-natgateway/azure_bdist_wheel.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+#--------------------------------------------------------------------------
+
+from distutils import log as logger
+import os.path
+
+from wheel.bdist_wheel import bdist_wheel
+class azure_bdist_wheel(bdist_wheel):
```
Comment (adewaleo) [inline src/command_modules/azure-cli-natgateway/azure_bdist_wheel.py]: @marstr and @tjprescott, if we end up consolidating the command modules, would we be able to remove redundant definitions of this class and instances of this file?

**Pull request merged**